torch/testing/_internal/compilable_quantized_tensor.py

```diff
@@ -0,0 +1,366 @@
+        (layout_type,) = tensor_attributes
+        return cls(int_data, scale, layout_type)
+
+    @classmethod
+    def from_plain(
+        cls,
+        int_data: torch.Tensor,
+        scale: torch.Tensor,
+        layout_type: LayoutType,
+    ) -> torch.Tensor:
+        """Construct a layout tensor from plain tensors and a layout_type, which main contain
+        extra metadata for packing etc.
+        """
+        assert isinstance(layout_type, PlainLayoutType)
+        return cls(int_data, scale, layout_type)
+
+    def _apply_fn_to_data(self, fn: Callable) -> torch.Tensor:
+        return self.__class__(
+            fn(self.int_data),
+            fn(self.scale),
+            self.layout_type,
+        )
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args, kwargs):  # type: ignore[no-untyped-def]
+        kwargs = {} if kwargs is None else kwargs
+
+        if func is aten.detach.default:
+            return return_and_correct_aliasing(
+                func, args, kwargs, args[0]._apply_fn_to_data(torch.detach)
+            )
+
+        raise NotImplementedError(
+            f"MyDTypeLayout dispatch: attempting to run {func}, this is not supported"
+        )
+
+
+#####################################################
+# torch functional and aten operator implementation #
+#####################################################
+
+implements = MyDTypeTensor.implements
+
+
+def _quantized_linear_op(
+    input_tensor: torch.Tensor, weight_tensor: torch.Tensor, bias: torch.Tensor
+) -> torch.Tensor:
+    if isinstance(input_tensor, MyDTypeTensor):
+        input_tensor = input_tensor.dequantize()
+    if isinstance(weight_tensor, MyDTypeTensor):
+        weight_tensor = weight_tensor.dequantize()
+    return torch.nn.functional.linear(input_tensor, weight_tensor, bias)
+
+
+@implements(torch.nn.functional.linear)
+def _(func, types, args, kwargs):  # type: ignore[no-untyped-def]
+    input_tensor, weight_tensor, bias = (
+        args[0],
+        args[1],
+        args[2] if len(args) > 2 else None,
+    )
+    # using try/except here so that we can have a general fallback when input_tensor/weight_tensor
+    # is not picked up by any of the dispatch paths in `_quantized_linear_op`, this allows us to
+    # make the branches easier to understand in `_quantized_linear_op`
+    try:
+        return _quantized_linear_op(input_tensor, weight_tensor, bias)
+    except NotImplementedError:
+        if isinstance(input_tensor, MyDTypeTensor):
+            input_tensor = input_tensor.dequantize()
+        if isinstance(weight_tensor, MyDTypeTensor):
+            weight_tensor = weight_tensor.dequantize()
+        return torch.nn.functional.linear(input_tensor, weight_tensor, bias)
+
+
+@implements(aten.detach.default)
+def _(func, types, args, kwargs):  # type: ignore[no-untyped-def]
+    # `return_and_correct_aliasing` should be used by wrapper tensor ``__torch_dispatch__`` subclasses that would like to
```

Comment:
fwiw, we want to drop this requirement as part of https://docs.google.com/document/d/1vpbGMsbdhPzFpXi8GJxzkzAGufs0slrIqKqfwDNwKj4/edit#heading=h.76wykr4y5vap (cc @IvanKobzarev )

Reply:
you mean Ivan is working on this?

Reply:
yep, it's on his list (we can keep the API in here for now though and remove it later)